### ISSUE TITLE
Docker demo for CV/NLP not working on n300.

### ISSUE THREAD

**Issue by changh95** [tenstorrent/tt-buda-demos#121]
**Describe the bug**

Demos in the 'first 5 steps' fail to run when run inside a docker container.
I'm using Wormhole n300. 

**To Reproduce**

1. Followed installation procedure in [first_5_steps/1_install_tt_buda.md](https://github.com/tenstorrent/tt-buda-demos/blob/main/first_5_steps/1_install_tt_buda.md). Made sure driver was installed, firmware was flashed to newest one, and hugepages were made.
2. Pull docker image ghcr.io/tenstorrent/tt-buda/ubuntu-22-04-amd64/wh_b0:v0.12.3
3. Run docker container with `sudo docker run --rm -ti --cap-add=sys_nice --shm-size=4g --device /dev/tenstorrent -v /dev/hugepages-1G:/dev/hugepages-1G -v $(pwd)/:/home/ ghcr.io/tenstorrent/tt-buda/ubuntu-22-04-amd64/wh_b0:v0.12.3 bash`
4. copy ipynb of nlp/cv demos into a python script.
5. Run the python script.

**Expected behavior**

Demo output should be shown.
Instead, I saw errors.

**System (please complete the following information):**
 - Device: n300
 - Host information: AMD Threadripper 2950x
 - OS: Ubuntu 24.04
 - Version: v0.12.3

**Additional context**

It's only the ipynb in 'first 5 steps' that do not work. When I tried the BERT demos inside `model_demos/nlp_demos/bert/...`, it worked nicely.

Error log:

```
2024-08-23 13:18:37.613 | WARNING  | generated_modules.pt_resnet50:process_framework_parameters:1985 - resnet.encoder.stages.3.layers.1.layer.2.normalization.num_batches_tracked not found in self._parameters and self._constants
2024-08-23 13:18:37.614 | WARNING  | generated_modules.pt_resnet50:process_framework_parameters:1985 - resnet.encoder.stages.3.layers.2.layer.0.normalization.num_batches_tracked not found in self._parameters and self._constants
2024-08-23 13:18:37.616 | WARNING  | generated_modules.pt_resnet50:process_framework_parameters:1985 - resnet.encoder.stages.3.layers.2.layer.1.normalization.num_batches_tracked not found in self._parameters and self._constants
2024-08-23 13:18:37.617 | WARNING  | generated_modules.pt_resnet50:process_framework_parameters:1985 - resnet.encoder.stages.3.layers.2.layer.2.normalization.num_batches_tracked not found in self._parameters and self._constants
2024-08-23 13:18:37.752 | DEBUG    | pybuda.ttdevice:_create_input_queue_device_connector:1408 - Creating input queue connector on TTDevice 'tt_device_0'
2024-08-23 13:18:37.752 | DEBUG    | pybuda.ttdevice:_create_intermediates_queue_device_connector:1418 - Creating fwd intermediates queue connector on TTDevice 'tt_device_0'
2024-08-23 13:18:37.752 | DEBUG    | pybuda.ttdevice:_create_forward_output_queue_device_connector:1398 - Creating forward output queue connector on TTDevice 'tt_device_0'
2024-08-23 13:18:37.753 | ERROR    | pybuda.run.impl:_start_device_processes:1180 - Process spawn error: 
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
2024-08-23 13:18:37.753 | DEBUG    | pybuda.run.impl:_shutdown:1265 - PyBuda shutdown
2024-08-23 13:18:37.753 | DEBUG    | pybuda.run.impl:_shutdown:1281 - Waiting until processes done
2024-08-23 13:18:37.854 | DEBUG    | pybuda.device:get_command_queue_response:311 - Ending process on TTDevice 'tt_device_0' due to shutdown event
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python3.10/multiprocessing/spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/usr/lib/python3.10/multiprocessing/spawn.py", line 125, in _main
    prepare(preparation_data)
  File "/usr/lib/python3.10/multiprocessing/spawn.py", line 236, in prepare
    _fixup_main_from_path(data['init_main_from_path'])
  File "/usr/lib/python3.10/multiprocessing/spawn.py", line 287, in _fixup_main_from_path
    main_content = runpy.run_path(main_path,
  File "/usr/lib/python3.10/runpy.py", line 289, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "/usr/lib/python3.10/runpy.py", line 96, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/test_cv_models.py", line 46, in <module>
    output_q = pybuda.run_inference()  # executes compilation (if first time) + runtime
  File "/usr/local/lib/python3.10/dist-packages/pybuda/run/api.py", line 90, in run_inference
    return _run_inference(module, inputs, input_count, output_queue, _sequential, _perf_trace, _verify_cfg)
  File "/usr/local/lib/python3.10/dist-packages/pybuda/run/impl.py", line 277, in _run_inference
    return _run_devices_inference(
  File "/usr/local/lib/python3.10/dist-packages/pybuda/run/impl.py", line 467, in _run_devices_inference
    output_queue = _initialize_pipeline(False, output_queue, sequential=sequential, verify_cfg=verify_cfg)
  File "/usr/local/lib/python3.10/dist-packages/pybuda/run/impl.py", line 414, in _initialize_pipeline
    _compile_devices(sequential, training=training, sample_inputs=sample_inputs, sample_targets=sample_targets, microbatch_count=microbatch_count, verify_cfg=verify_cfg)
  File "/usr/local/lib/python3.10/dist-packages/pybuda/run/impl.py", line 1250, in _compile_devices
    raise RuntimeError(f"Compile failed for {d}")
RuntimeError: Compile failed for TTDevice 'tt_device_0'
2024-08-23 13:18:37.855 | DEBUG    | pybuda.run.impl:_shutdown:1265 - PyBuda shutdown
2024-08-23 13:18:37.855 | DEBUG    | pybuda.run.impl:_shutdown:1281 - Waiting until processes done
Exception ignored in atexit callback: <function shutdown at 0x7f2f0598f400>
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/pybuda/run/api.py", line 475, in shutdown
    return _shutdown()
  File "/usr/local/lib/python3.10/dist-packages/pybuda/run/impl.py", line 1294, in _shutdown
    p.terminate()
  File "/usr/lib/python3.10/multiprocessing/process.py", line 133, in terminate
    self._popen.terminate()
AttributeError: 'NoneType' object has no attribute 'terminate'
^CTraceback (most recent call last):
  File "/home/test_cv_models.py", line 46, in <module>
    output_q = pybuda.run_inference()  # executes compilation (if first time) + runtime
  File "/usr/local/lib/python3.10/dist-packages/pybuda/run/api.py", line 90, in run_inference
    return _run_inference(module, inputs, input_count, output_queue, _sequential, _perf_trace, _verify_cfg)
  File "/usr/local/lib/python3.10/dist-packages/pybuda/run/impl.py", line 277, in _run_inference
    return _run_devices_inference(
  File "/usr/local/lib/python3.10/dist-packages/pybuda/run/impl.py", line 467, in _run_devices_inference
    output_queue = _initialize_pipeline(False, output_queue, sequential=sequential, verify_cfg=verify_cfg)
  File "/usr/local/lib/python3.10/dist-packages/pybuda/run/impl.py", line 411, in _initialize_pipeline
    ctx.processes = _start_device_processes(devices, _get_global_compiler_config().backend_output_dir)
  File "/usr/local/lib/python3.10/dist-packages/pybuda/run/impl.py", line 1178, in _start_device_processes
    p.start()
  File "/usr/lib/python3.10/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/usr/lib/python3.10/multiprocessing/context.py", line 288, in _Popen
    return Popen(process_obj)
  File "/usr/lib/python3.10/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/usr/lib/python3.10/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/usr/lib/python3.10/multiprocessing/popen_spawn_posix.py", line 62, in _launch
    f.write(fp.getbuffer())
KeyboardInterrupt
```

**Comment by milank94** [comment]
@changh95 if you converted the `.ipynb` files to `.py`, please ensure to also wrap the functions in a `if __name__ == '__main__':` header. That will resolve the spawn process error.

**Comment by changh95** [comment]
@milank94
I added the `if __name__ == '__main__':` and the code still fails to run.
I'm getting Platform error - it seems like TT cannot find my chip?
I'm using n300, so there are 2 chips.

2024-09-06 11:55:39.187 | INFO     | Runtime         - running: '/usr/local/lib/python3.10/dist-packages/budabackend/umd/device/bin/silicon/x86/create-ethernet-map /usr/local/lib/python3.10/dist-p
ackages/budabackend//cluster_desc.yaml'                                                           
Error: PlatformError(Generic("When detecting chips in mesh found a mismatch between the expected chip coordinate [0, 0, 1, 0] and the actual [0, 0, 0, 0]", BtWrapper(<disabled>)))
2024-09-06 11:55:39.221 | FATAL    | Always          - Unable to generate cluster descriptor file 
terminate called after throwing an instance of 'std::runtime_error'                               
  what():  Unable to generate cluster descriptor file                                             
backtrace:                                                                                        
 --- /usr/local/lib/python3.10/dist-packages/pybuda/../budabackend/build/lib/libtt.so(+0x20d3c8) [0x78a382a183c8]
 --- tt_cluster::get_cluster_desc_path(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
 --- tt_cluster::detect_available_devices(tt::TargetDevice const&, bool)                          
 --- tt::param::get_device_descs(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
 --- tt::backend::get_device_descs_for_available_devices(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
 --- tt::query_available_tt_devices()                                                             
 --- tt::TorchDeviceImpl::get()         
 --- /usr/local/lib/python3.10/dist-packages/pybuda/_C.cpython-310-x86_64-linux-gnu.so(+0x27b880) [0x78a3834d2880]
 --- /lib64/ld-linux-x86-64.so.2(+0x647e) [0x78a47da4a47e]                                        
 --- /lib64/ld-linux-x86-64.so.2(+0x6568) [0x78a47da4a568]    
 --- /lib/x86_64-linux-gnu/libc.so.6(_dl_catch_exception+0xe5) [0x78a47d851af5]
 --- /lib64/ld-linux-x86-64.so.2(+0xdff6) [0x78a47da51ff6]                     
 --- /lib/x86_64-linux-gnu/libc.so.6(_dl_catch_exception+0x88) [0x78a47d851a98]
 --- /lib64/ld-linux-x86-64.so.2(+0xe34e) [0x78a47da5234e]                     
 --- /lib/x86_64-linux-gnu/libc.so.6(+0x9063c) [0x78a47d76d63c]
 --- /lib/x86_64-linux-gnu/libc.so.6(_dl_catch_exception+0x88) [0x78a47d851a98]
 --- /lib/x86_64-linux-gnu/libc.so.6(_dl_catch_error+0x33) [0x78a47d851b63]    
 --- /lib/x86_64-linux-gnu/libc.so.6(+0x9012e) [0x78a47d76d12e]                                   
 --- /lib/x86_64-linux-gnu/libc.so.6(dlopen+0x48) [0x78a47d76d6c8]
 --- python3(+0x2614fb) [0x563d6b5a04fb]                                                          
 --- python3(+0x25fa27) [0x563d6b59ea27] 

---

Here's screenshot of my tt-smi.

![image](https://github.com/user-attachments/assets/3a4683d7-c2ec-4564-8457-0d853edf7ecb)

**Comment by changh95** [comment]
After rebooting, the error disappeared (it seems like TT could find my chip).
But the code fails, as it tries to use more memory than what the card can afford.
`update_buffer: Buffer OVERLAY_BLOB is using more l1 memory than reserved (66196 > 65408)`

There should be plenty L1 memory in the card (at least if the demo is designed for n300). Any idea why the code fails?

# Error log

2024-09-06 12:28:06.186 | FATAL    | Always          - update_buffer: Buffer OVERLAY_BLOB is using more l1 memory than reserved (66196 > 65408)
2024-09-06 12:28:06.195 | INFO     | Runtime         - Compile result: FAILURE, failure_type: undefined, device_id: 0, temporal_epoch_id: 0, graph: , logical_core_x: 0, logical_core_y: 0, failure_
target:                                                                                           
2024-09-06 12:28:06.195 | ERROR    | Always          - update_buffer: Buffer OVERLAY_BLOB is using more l1 memory than reserved (66196 > 65408)
backtrace:                                                                                        
 --- perf::L1Core::update_buffer(perf::L1BufferType const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned int, unsigned int, int, int)
 --- perf::L1Graph::update_buffer_of_core(tt::umd::cxy_pair const&, perf::L1BufferType const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned int,
 unsigned int, int, int, perf::CoordType)                                                                                                                                                           
 --- perf::L1Profiler::update_buffer_of_graph(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, tt::umd::cxy_pair const&, perf::L1BufferType const&, std::__cx
x11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned int, unsigned int, int, int, perf::CoordType, perf::L1ProfileStage)
 --- perf::MemoryProfiler::update_buffer_of_graph_l1(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, tt::umd::cxy_pair const&, perf::L1BufferType const&, st
d::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned int, unsigned int, int, int, perf::CoordType, perf::L1ProfileStage)
 --- tt_runtime::profile_actual_l1_binary_buffers()
 --- tt_runtime::initialize(tt::tt_compile_result*)                                                                                                                                                 
 --- /usr/local/lib/python3.10/dist-packages/pybuda/_C.cpython-310-x86_64-linux-gnu.so(+0x8ad434) [0x6ffd53d04434]
 --- /usr/local/lib/python3.10/dist-packages/pybuda/_C.cpython-310-x86_64-linux-gnu.so(+0x2bf286) [0x6ffd53716286]
 --- /usr/bin/python3(+0x15a10e) [0x57cc86ac410e]                 
 --- /usr/bin/python3(_PyObject_MakeTpCall+0x25b) [0x57cc86abaa7b]
 --- /usr/bin/python3(+0x168acb) [0x57cc86ad2acb]                      
 --- /usr/bin/python3(_PyEval_EvalFrameDefault+0x614a) [0x57cc86ab2cfa]
 --- /usr/bin/python3(_PyObject_FastCallDictTstate+0xc4) [0x57cc86ab9c14]
 --- /usr/bin/python3(+0x164b05) [0x57cc86aceb05]                        
 --- /usr/bin/python3(_PyObject_MakeTpCall+0x1fc) [0x57cc86abaa1c]
 --- /usr/bin/python3(_PyEval_EvalFrameDefault+0x64e6) [0x57cc86ab3096]
 --- /usr/bin/python3(_PyFunction_Vectorcall+0x7c) [0x57cc86ac49fc]    
 --- /usr/bin/python3(_PyEval_EvalFrameDefault+0x8ac) [0x57cc86aad45c]
 --- /usr/bin/python3(_PyFunction_Vectorcall+0x7c) [0x57cc86ac49fc]   
 --- /usr/bin/python3(_PyEval_EvalFrameDefault+0x8ac) [0x57cc86aad45c]
 --- /usr/bin/python3(+0x16893e) [0x57cc86ad293e]                     
 --- /usr/bin/python3(_PyEval_EvalFrameDefault+0x2a27) [0x57cc86aaf5d7]
 --- /usr/bin/python3(_PyFunction_Vectorcall+0x7c) [0x57cc86ac49fc]    
 --- /usr/bin/python3(_PyEval_EvalFrameDefault+0x8ac) [0x57cc86aad45c]
 --- /usr/bin/python3(_PyFunction_Vectorcall+0x7c) [0x57cc86ac49fc]   
 --- /usr/bin/python3(_PyEval_EvalFrameDefault+0x8ac) [0x57cc86aad45c]
 --- /usr/bin/python3(_PyFunction_Vectorcall+0x7c) [0x57cc86ac49fc]   
 --- /usr/bin/python3(_PyEval_EvalFrameDefault+0x6bd) [0x57cc86aad26d]
 --- /usr/bin/python3(_PyFunction_Vectorcall+0x7c) [0x57cc86ac49fc]   
 --- /usr/bin/python3(_PyEval_EvalFrameDefault+0x198c) [0x57cc86aae53c]
 --- /usr/bin/python3(+0x13f9c6) [0x57cc86aa99c6]                      
 --- /usr/bin/python3(PyEval_EvalCode+0x86) [0x57cc86b9f256]
 --- /usr/bin/python3(+0x260108) [0x57cc86bca108]           
 --- /usr/bin/python3(+0x2599cb) [0x57cc86bc39cb] 
 --- /usr/bin/python3(PyRun_StringFlags+0x81) [0x57cc86bbcab1]
 --- /usr/bin/python3(PyRun_SimpleStringFlags+0x41) [0x57cc86bbc961]
 --- /usr/bin/python3(Py_RunMain+0x375) [0x57cc86bbbb15]            
 --- /usr/bin/python3(Py_BytesMain+0x2d) [0x57cc86b9202d]
 --- /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x6ffe4d9bad90]
 --- /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80) [0x6ffe4d9bae40]
 --- /usr/bin/python3(_start+0x25) [0x57cc86b91f25]                          
                                                                                                  
2024-09-06 12:28:06.195 | INFO     | Runtime         - Cleaning up runtime and closing device.
2024-09-06 12:28:06.195 | INFO     | Runtime         - Waiting for cluster completion
2024-09-06 12:28:06.196 | INFO     | PerfPostProcess - Writing the host postprocess report in /tmp/root/94aeb69a5b4e/perf_results//host/device_alignment_th_4223468125_proc_511.json
2024-09-06 12:28:06.284 | INFO     | Runtime         - Finished clean up successfully                                                                                                               
2024-09-06 12:28:06.285 | INFO     | pybuda.backend:__init__:85 - Backend compile False, target: , error type: BackendCompileFailure.Invalid, error: update_buffer: Buffer OVERLAY_BLOB is using mor
e l1 memory than reserved (66196 > 65408)                                                         
backtrace:

**Comment by milank94** [comment]
@changh95 let's try a few debug steps.

1. Can you use the latest version of `pybuda` - v0.18.2: https://github.com/tenstorrent/tt-buda/pkgs/container/tt-buda%2Fubuntu-22-04-amd64%2Fwh_b0/245601102?tag=v0.18.2
2. Is this the latest version of the `.ipynb` script you are trying to convert and run? https://github.com/tenstorrent/tt-buda-demos/blob/main/first_5_steps/2_running_nlp_models.ipynb
3. By default, the demo runs on a single-chip (only one of the chips in the n300). To run on dual-chip (data parallel mode only), you need the two following modifications:
    - include the env variable `PYBUDA_N300_DATA_PARALLEL=1`
    - the input batch size must be great than 2

**Comment by changh95** [comment]
Thank you for prompt reply.

I upgraded my docker version to v0.18.2, and everything worked flawlessly. 

One thing I noticed is that the documentation for setting up tt-buda ([first_5_steps/1_install_tt_buda.md](https://github.com/tenstorrent/tt-buda-demos/blob/main/first_5_steps/1_install_tt_buda.md)) may mislead the first-time users to install v0.12.3. Below screenshot is a snippet from the documentation, and by saying `version number i.e. v0.12.3` it would mean that the version number IS v0.12.3. Many first time users would assume the documentation is up-to-date, hence the latest working version. I myself was not aware the latest version was this high up!

I would suggest updating the docs as `version number e.g. v0.18.2`? So even the first time users can know it's just a placeholder for version numbers, and v0.18.2 is one of them. Maybe we can add a link that takes us to the latest release of tt-buda.

![image](https://github.com/user-attachments/assets/46e49864-d2ea-4ab3-a1cf-dcdb367af9a6)
